feat: Add workflow to check config description

PeterChen13579 opened the pull request:
fixes #1880

PeterChen13579 committed:
add new workflow

PeterChen13579 committed:
fix

kuznetsss committed:
ci: Upload cache only for develop branch (#1897)

PeterChen13579 committed:
test: add non-admin test for simulate (#1893)

PeterChen13579 committed:
fix path

PeterChen13579 committed:
fix syntax

PeterChen13579 committed:
fix value

cindyyan317 commented on .github/workflows/build.yml:
```diff
@@ -127,3 +127,44 @@ jobs:
+          repoConfigHash=$(sha256sum ${repoConfigFile} | cut -d' ' -f1)
+
+          if [ ${configDescriptionHash} != ${repoConfigHash} ]; then
+            diff -u "${repoConfigFile}" "${configDescriptionFile}"
```
Maybe add error message like "markdown file not up to date"

cindyyan317 commented on .github/workflows/build.yml:
```diff
@@ -127,3 +127,44 @@ jobs:
+      - name: Compare Config Description
+        shell: bash
+        run: |
+          chmod +x ./clio_server
```
We can run this line after `repoConfigFile` check.

cindyyan317 reviewed (commented):
Logic looks good to me. 

I am not against adding generated file to repo. I think it is better to give developer enough instruction how to update this generated file. Please ignore my comment if there is instruction somewhere 😄 
For example: we can
1 Give instruction when check fails in CI job
2 Create a CMake target which depends on ConfigDescription.cpp file, it will update the markdown file during building when `ConfigDescription.cpp` is amended.
3 At the beginning of the config-description.md, we add `This is generated file, don't amend it manually , please run balabala`

kuznetsss committed:
fix: Array parsing in new config (#1896)

Improving array parsing in config:
- Allow null values in arrays for optional fields
- Allow empty array even for required field
- Allow to not put an empty array in config even if array contains
required fields

kuznetsss committed:
fix: Better errors on logger init failure (#1857)

Fixes #1326.

PeterChen13579 commented on docs/config-description.md:
```diff
@@ -0,0 +1,452 @@
+### Key: database.cassandra.threads
+- **Required**: True
+- **Type**: int
+- **Default value**: 32
```
@godexsoft , @kuznetsss and I discussed, because the default value is generated by `thread::hardware_concurrency()` the value of .md file that is produced locally vs on github is different 🫤

cindyyan317 replied:
In this case, how user should respond this checking failure?

PeterChen13579 replied:
It's fixed by adding description into default value 😄  https://github.com/XRPLF/clio/pull/1894/files#diff-55af6dfc7ab80d735efe30bc7f4edc67087324425d9247db7405162f0a7e4f9dR279

PeterChen13579 committed:
fix comments

godexsoft committed:
feat: ETLng monitor (#1898)

For #1594

github-actions[bot] committed:
style: clang-tidy auto fixes (#1920)

Fixes #1919. Please review and commit clang-tidy fixes.

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

shawnxie999 committed:
feat: LPT freeze (#1840)

Fixes #1827

github-actions[bot] committed:
style: clang-tidy auto fixes (#1925)

Fixes #1924. Please review and commit clang-tidy fixes.

Co-authored-by: godexsoft <385326+godexsoft@users.noreply.github.com>

kuznetsss committed:
fix: Fix backtrace usage (#1932)

kuznetsss committed:
fix: Data race in new webserver (#1926)

There was a data race inside `CoroutineGroup` because internal timer was
used from multiple threads in the methods `asyncWait()` and
`onCoroutineComplete()`. Changing `registerForeign()` to spawn to the
same `yield_context` fixes the problem because now the timer is accessed
only from the same coroutine which has an internal strand.

During debugging I also added websocket support for `request_gun` tool.

PeterChen13579 committed:
Merge branch 'develop' into addConfigDescriptionToCI

PeterChen13579 committed:
fix hardware_concurrency

PeterChen13579 committed:
Delete docs/config-description.md

PeterChen13579 committed:
Rename config_description.md to config-description.md

kuznetsss commented on src/util/newconfig/ConfigValue.hpp:
```diff
@@ -213,7 +215,11 @@ class ConfigValue {
         stream << "- **Required**: " << (val.isOptional() ? "False" : "True") << "\n";
         stream << "- **Type**: " << val.type() << "\n";
-        stream << "- **Default value**: " << (val.hasValue() ? *val.value_ : "None") << "\n";
+        if (val.hasValue() && val.description_.has_value()) {
```
```suggestion
        if (val.description_.has_value()) {
```

kuznetsss commented on src/util/newconfig/ConfigDefinition.hpp:
```diff
@@ -274,7 +274,10 @@ static ClioConfigDefinition gClioConfig = ClioConfigDefinition{
-          .defaultValue(static_cast<uint32_t>(std::thread::hardware_concurrency()))
+          .defaultValue(
+              static_cast<uint32_t>(std::thread::hardware_concurrency()),
+              "The number of threads in your hardware system."
```
Single CPU could run multiple threads.
```suggestion
              "The number of available CPU cores."
```

kuznetsss commented on src/util/newconfig/ConfigDefinition.hpp:
```diff
@@ -313,8 +316,9 @@ static ClioConfigDefinition gClioConfig = ClioConfigDefinition{
-      ConfigValue{ConfigType::Integer}.defaultValue(std::thread::hardware_concurrency()).withConstraint(gValidateUint32)
-     },
+      ConfigValue{ConfigType::Integer}
+          .defaultValue(std::thread::hardware_concurrency(), "The number of threads in your hardware system.")
```
```suggestion
          .defaultValue(std::thread::hardware_concurrency(), "The number of available CPU cores.")
```

PeterChen13579 committed:
fix comments

kuznetsss reviewed (approved):
👍
